### PR TITLE
Feature/templates

### PR DESCRIPTION
**Pull request by olejiksin**
creating new file from document editor by document template

### CODE REVIEW

**Review by Eloncase — APPROVED** [review]

File: repo/src/main/java/com/parashift/onlyoffice/Prepare.java [hunk 1]
```diff
@@ -102,8 +104,17 @@ public void execute(WebScriptRequest request, WebScriptResponse response) throws
                         if (pathLocale == null) pathLocale = Util.PathLocale.get("en");
                     }
 
-                    InputStream in = getClass().getResourceAsStream("/newdocs/" + pathLocale + "/new." + ext);
-
+                    InputStream in;
+                    NodeRef parentNodeRef = null;
+                    if(request.getParameter("parentNodeRef") != null){
+                        parentNodeRef = new NodeRef(request.getParameter("parentNodeRef"));
+                        URL url = new URL(util.getContentUrl(parentNodeRef));
```
Comment (aleksandrfedorov97) [inline repo/src/main/java/com/parashift/onlyoffice/Prepare.java:111]: get inputStream using contentService.getReader()

File: share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.html.ftl [hunk 2]
```diff
@@ -29,6 +29,23 @@
...
+                            image: "${share}proxy/alfresco/api/node/workspace/SpacesStore/" + node.nodeRef.split("/SpacesStore/")[1] +"/content/thumbnails/doclib?ph=true",
+                            title: node.name,
+                            url: config.editorConfig.createUrl + "&parentNodeRef=" + node.nodeRef
+                        };
+                        templates.push(template);
+                    }
+                }
+            config.editorConfig.templates = templates;
+        });
+
```
Comment (aleksandrfedorov97) [inline share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.html.ftl:48]: is it possible to get an templatings object on the server side?

**Review by aleksandrfedorov97 — CHANGES_REQUESTED** [review]

**Review by aleksandrfedorov97 — APPROVED** [review]

File: share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.js [hunk 3]
```diff
@@ -6,6 +6,7 @@
 var query = "?nodeRef=" + url.args.nodeRef;
 if (url.args.readonly) query += "&readonly=1";
 if (url.args.new) query += "&new=" + url.args.new;
+if (url.args.parentNodeRef) query+= "&parentNodeRef=" + url.args.parentNodeRef;
```
Comment (LinneyS) [inline share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.js:9]: encodeURIComponent

File: share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.js [hunk 4]
```diff
@@ -6,6 +6,7 @@
 var query = "?nodeRef=" + url.args.nodeRef;
```
Comment (LinneyS) [inline share/src/main/resources/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.js:6]: encodeURIComponent

File: repo/src/main/java/com/parashift/onlyoffice/Util.java [hunk 5]
```diff
@@ -180,6 +193,38 @@ public String generateHash() {
...
+        JSONArray templates = new JSONArray();
+        NodeRef templatesNodeRef = getNodeByPath("/app:company_home/app:dictionary/app:node_templates");
+        List<ChildAssociationRef> assocs = nodeService.getChildAssocs(templatesNodeRef);
+        for(ChildAssociationRef assoc : assocs){
+            String docName = nodeService.getProperty(assoc.getChildRef(), ContentModel.PROP_NAME).toString();
+            if(docExt.equals(docName.substring(docName.lastIndexOf(".") + 1))){
+                JSONObject template = new JSONObject();
+                String image = getShareUrl() + "proxy/alfresco/api/node/workspace/SpacesStore/" + assoc.getChildRef().toString().split("/SpacesStore/")[1] + "/content/thumbnails/doclib?ph=true";
+                String title = nodeService.getProperty(assoc.getChildRef(), ContentModel.PROP_NAME).toString();
+                String url = getCreateNewUrl(nodeRef, mimeType) + "&parentNodeRef=" + assoc.getChildRef();
```
Comment (LinneyS) [inline repo/src/main/java/com/parashift/onlyoffice/Util.java:214]: encode

**Review by LinneyS — CHANGES_REQUESTED** [review]

**Pull request merged**